### PR TITLE
Example code for blog on new row comparators

### PR DESCRIPTION
**Pull request by divyegala**
## Description
Example code using a few libcudf APIs to demonstrate nested-type usage.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by karthikeyann** [comment]
@divyegala which is a good location for adding code to round tripping data using JSON reader and writer?

**Comment by karthikeyann** [comment]
Added metadata round trip for json reader, writer usages.

**Comment by divyegala** [comment]
@karthikeyann can you rebase and remove your commits? I removed `example.cpp` and only have `deduplication.cpp` in my working branch locally. When I try to merge your commits and run the examples, I get `std::bad_alloc()` and I'm probably doing a bad merge somewhere. It'll be easier for me to push my commits and for you to apply yours on top of mine

**Comment by karthikeyann** [comment]
The cmake compilation didn’t work right away.
Added `-DCMAKE_CUDA_ARCHITECTURES=native`  to all cmake commands in cpp/examples/build.sh to compile.

**Comment by GregoryKimball** [comment]
@divye Would you please update this example to use an RMM pool? @davidwendt added this in `examples/strings/common.hpp`

File: cpp/examples/nested_types/deduplication.cpp
Comment (GregoryKimball) [inline cpp/examples/nested_types/deduplication.cpp:139]:
```suggestion
  if (argc < 4) {
    input_filepath  = "example.json";
    output_filepath = "output.json";
    mr_name         = "pool";
  } else if (argc == 4) {
```

**Comment by GregoryKimball** [comment]
@vyasr and @davidwendt would you please share your feedback on the way this example demonstrates the libcudf public API?

File: cpp/examples/nested_types/deduplication.cpp
Comment (PointKernel) [inline cpp/examples/nested_types/deduplication.cpp:52]:
```suggestion
std::shared_ptr<rmm::mr::device_memory_resource> create_memory_resource(std::string_view const name)
```
nit: `string_view` instead of `string &`

File: cpp/examples/nested_types/deduplication.cpp
Comment (PointKernel) [inline cpp/examples/nested_types/deduplication.cpp:35]: I find it useful to have one or two paragraphs summarizing what we want to demonstrate in the current example, e.g. https://github.com/NVIDIA/cuCollections/blob/dev/examples/static_map/device_view_example.cu#L31-L42

File: cpp/examples/nested_types/deduplication.cpp
Comment (PointKernel) [inline cpp/examples/nested_types/deduplication.cpp:72]:
```suggestion
  auto builder  = cudf::io::json_writer_options::builder(sink_info, tbl).lines(true);
  builder.metadata(metadata);
  auto options = builder.build();
```
Not a hard requirement: there are no naming conflicts between this and the above `read_json` function thus `builder` and `options` are good enough IMO. If we want more descriptive names: could use `writer_builder` and `writer_options` instead.

File: cpp/examples/nested_types/deduplication.cpp
Comment (PointKernel) [inline cpp/examples/nested_types/deduplication.cpp:66]:
```suggestion
void write_json(cudf::table_view input, cudf::io::table_metadata input_metadata, std::string filepath)
```
Not saying the suggestions are better. Just throwing ideas to use more descriptive names and avoid abbreviations.

File: cpp/examples/nested_types/deduplication.cpp
Comment (PointKernel) [inline cpp/examples/nested_types/deduplication.cpp:154]:
```suggestion
  auto const start_time = std::chrono::steady_clock::now();
```

File: cpp/examples/nested_types/deduplication.cpp
Comment (PointKernel) [inline cpp/examples/nested_types/deduplication.cpp:158]: The example output would look like this:
```bash
Wall time: 1.0 seconds
Wall time: 2.0 seconds
Wall time: 3.0 seconds
Wall time: 4.0 seconds
```
Could we print out more specific timing info? something like (just the rough idea, you see what I mean):
```bash
Count aggregation time: 1.0 seconds
Join time: 2.0 seconds
Sort time: ...
```

**Review by PointKernel — COMMENTED** [review]
Some minor suggestions.

Probably worth adding doxygen docs for all local functions as well.

File: cpp/examples/nested_types/deduplication.cpp
Comment (davidwendt) [inline cpp/examples/nested_types/deduplication.cpp:94]: I know this may be out of scope for this specific example but I've always found the groupby API logic very difficult to follow. I understand the current API pattern may provide the most flexibility but I don't see how someone could come up with this logic just from the documentation. Since the documentation would not be able to cover every possible aggregation, I think it would serve libcudf users better by developing a good set of groupby examples.

Comment (divyegala) [reply cpp/examples/nested_types/deduplication.cpp:94]: Thanks @davidwendt, I entirely agree with your sentiment here. Even after writing the example I still could not write a groupby aggregation from memory. We went through iterations of different aggravations while developing this example and every time I would copy-paste groupby code.

Comment (davidwendt) [reply cpp/examples/nested_types/deduplication.cpp:94]: groupby-aggregations == aggravations 😆

Comment (divyegala) [reply cpp/examples/nested_types/deduplication.cpp:94]: My phone autocorrect knows what's up!

File: cpp/examples/nested_types/deduplication.cpp
Comment (GregoryKimball) [reply cpp/examples/nested_types/deduplication.cpp:159]: One comment from the console parsing side, using the pattern: 
`{func_name} time: X seconds` is easier to parse on a `.split(' ')`

File: cpp/examples/nested_types/deduplication.cpp
Comment (GregoryKimball) [inline cpp/examples/nested_types/deduplication.cpp:185]:
```suggestion
  auto count_w_copy    = count_aggregate_with_copy(tbl->view());
```

File: cpp/examples/nested_types/deduplication.cpp
Comment (divyegala) [reply cpp/examples/nested_types/deduplication.cpp:67]: As we are using `tbl` abbrevation everywhere for `table_view` type, I'd like to keep it as that

File: cpp/examples/nested_types/deduplication.cpp
Comment (bdice) [inline cpp/examples/nested_types/deduplication.cpp:76]: Use a `string_view` instead of a `string`.
```suggestion
cudf::io::table_with_metadata read_json(std::string_view filepath)
```

File: cpp/examples/nested_types/CMakeLists.txt
Comment (bdice) [inline cpp/examples/nested_types/CMakeLists.txt:19]:
```suggestion
set(CUDF_TAG branch-23.12)
```

File: cpp/examples/nested_types/deduplication.cpp
Comment (bdice) [inline cpp/examples/nested_types/deduplication.cpp:45]: Don't indent this. Match item 3.
```suggestion
 * 2. equality: Used by functions `count_aggregate` and `join_count` in conjunction with hashing
 * to determine equality for nested types
```

File: cpp/examples/nested_types/deduplication.cpp
Comment (bdice) [inline cpp/examples/nested_types/deduplication.cpp:50]: Should we annotate any code with comments explaining how to do it fast? It can be confusing when examples don't show best practices, especially for a performance-focused library.

File: cpp/examples/nested_types/deduplication.cpp
Comment (bdice) [inline cpp/examples/nested_types/deduplication.cpp:127]:
```suggestion
  auto new_left  = cudf::gather(left, cudf::device_span<cudf::size_type const>{*left_indices});
  auto new_right = cudf::gather(right, cudf::device_span<cudf::size_type const>{*right_indices});
```

File: cpp/examples/nested_types/deduplication.cpp
Comment (bdice) [inline cpp/examples/nested_types/deduplication.cpp:101]: Do count aggregations return `int32_t` or `size_type`? I think it's supposed to be `size_type` but not 100% sure.
```suggestion
  auto val  = cudf::make_numeric_column(cudf::data_type{cudf::type_to_id<cudf::size_type>()}, keys.num_rows());
```

File: cpp/examples/nested_types/deduplication.cpp
Comment (bdice) [reply cpp/examples/nested_types/deduplication.cpp:84]: I'm not sure what you mean by "everywhere" but we're pretty consistent about naming `table_view` objects `input` in libcudf. I would adopt @PointKernel's suggestion throughout.

File: cpp/examples/nested_types/deduplication.cpp
Comment (bdice) [inline cpp/examples/nested_types/deduplication.cpp:184]: Is showing timing results helpful when the example input is so small?

File: cpp/include/cudf/detail/gather.cuh
Comment (bdice) [inline cpp/include/cudf/detail/gather.cuh:486]: Quite a few of the changes in this PR (here and below) don't belong in a PR aimed at introducing new example code. They should be fixed in a separate PR.

File: cpp/examples/nested_types/deduplication.cpp
Comment (divyegala) [reply cpp/examples/nested_types/deduplication.cpp:184]: The example input is just an example, after all. You could use your own JSON input file with this.

File: cpp/examples/nested_types/deduplication.cpp
Comment (divyegala) [reply cpp/examples/nested_types/deduplication.cpp:84]: Everywhere in this example. Will change it.

File: cpp/examples/nested_types/deduplication.cpp
Comment (divyegala) [reply cpp/examples/nested_types/deduplication.cpp:50]: I don't think any of the code in this example is not the most-performant way to run that algorithm in isolation. However, a user could theoretically find other ways to pipeline this example better based on their input and requirements, and a disclaimer helps cover that scenario.

File: cpp/examples/nested_types/deduplication.cpp
Comment (divyegala) [reply cpp/examples/nested_types/deduplication.cpp:101]: It's just a dummy column to force usage of hash aggregation instead of sort aggregation, it does not hold the return values.

File: cpp/examples/nested_types/deduplication.cpp
Comment (ttnghia) [reply cpp/examples/nested_types/deduplication.cpp:84]: Our (implicit) convention in cudf is to use `input` "everywhere". My convention/recommendation is to use `_tview` or `_tv` suffix for `table_view` variables, but only for local variables. APIs' input parameters should (almost) always be `input`.

File: cpp/include/cudf/detail/gather.cuh
Comment (ttnghia) [reply cpp/include/cudf/detail/gather.cuh:486]: So if this PR requires such fix to work, please post a fix PR and mark as dependency of this one.

File: cpp/examples/nested_types/deduplication.cpp
Comment (bdice) [reply cpp/examples/nested_types/deduplication.cpp:50]: I would just drop this disclaimer. Otherwise it implies that there's a _known_ better way, which leaves me confused as a reader.

File: cpp/examples/nested_types/deduplication.cpp
Comment (bdice) [reply cpp/examples/nested_types/deduplication.cpp:101]: I'm not sure if I understand. We usually want to rely on hash aggregation anytime we can, since it tends to be faster than sort aggregations. Can you point to the behavior that you're avoiding, and why this is the right solution (as opposed to fixing the aggregation dispatch to use hashing instead of sorting)?

File: cpp/examples/nested_types/deduplication.cpp
Comment (bdice) [reply cpp/examples/nested_types/deduplication.cpp:184]: I agree, but I think it's unlikely that a user will want to use this exact pipeline. In other words, it's not helpful to show timing if the provided input example is so small. I think you have two choices:
1. just execute the pipeline without showing timing, since you're mostly just measuring launch latency.
2. write a script that generates a sufficiently large input file (where the timing becomes a meaningful measure of speed) and ship it with this example.

File: cpp/examples/nested_types/deduplication.cpp
Comment (divyegala) [reply cpp/examples/nested_types/deduplication.cpp:101]: https://github.com/rapidsai/cudf/blob/abc0d41d1d9033d581948ae19384e0aa0f33da77/cpp/src/groupby/hash/groupby.cu#L654-L656

Comment (divyegala) [reply cpp/examples/nested_types/deduplication.cpp:101]: This isn't the right solution. It should definitely be fixed at the core of the problem.

File: cpp/examples/nested_types/deduplication.cpp
Comment (divyegala) [reply cpp/examples/nested_types/deduplication.cpp:184]: @GregoryKimball should we ship your script that generates large inputs?

File: cpp/examples/nested_types/deduplication.cpp
Comment (bdice) [reply cpp/examples/nested_types/deduplication.cpp:101]: Is this something we now have the machinery to fix? I'm not sure what stops this from working. We should have full nested type support now.

File: cpp/examples/nested_types/deduplication.cpp
Comment (vyasr) [inline cpp/examples/nested_types/deduplication.cpp:141]: suggestion (non-blocking): Perhaps it's worth adding a comment that if stability is desired you could use `stable_sorted_order` here. Someone is always bound to want that and including information about it in more places seems helpful.

File: cpp/examples/nested_types/deduplication.cpp
Comment (vyasr) [inline cpp/examples/nested_types/deduplication.cpp:151]: suggestion (blocking): let's include a sample command-line invocation to help users out.

File: cpp/examples/nested_types/deduplication.cpp
Comment (vyasr) [inline cpp/examples/nested_types/deduplication.cpp:171]: suggestion (non-blocking): this error isn't quite doing what you want. It's probably better to be explicit here.

```suggestion
  std::string input_filepath;
  std::string output_filepath;
  std::string mr_name;
  if (argc != 4 && argc != 1) {
    std::cout << "Either provide all command-line arguments, or none to use defaults" << std::endl;
    return 1;
  }
  if (argc == 1) {
    input_filepath  = "example.json";
    output_filepath = "output.json";
    mr_name         = "pool";
  } else {
    input_filepath  = argv[1];
    output_filepath = argv[2];
    mr_name         = argv[3];
  }
```

File: cpp/examples/nested_types/deduplication.cpp
Comment (vyasr) [inline cpp/examples/nested_types/deduplication.cpp:181]: nitpick: This seems redundant. Can we just use `count_st` as the overall start when computing the total elapsed time at the end?

File: cpp/examples/nested_types/deduplication.cpp
Comment (vyasr) [inline cpp/examples/nested_types/deduplication.cpp:183]: nitpick: Can this be auto-deduced, or will the operator result in lower precision by default or something like that?

File: cpp/examples/nested_types/deduplication.cpp
Comment (vyasr) [reply cpp/examples/nested_types/deduplication.cpp:184]: thought: I'm inclined to say we should drop the timings. The performance of this example is not itself going to be valuable.

File: cpp/include/cudf/detail/gather.cuh
Comment (vyasr) [reply cpp/include/cudf/detail/gather.cuh:486]: suggestion (non-blocking): Ideally I'd also split the nvtx range changes from the nullable->has_nulls changes since they're fairly different classes of issues.

**Review by vyasr — CHANGES_REQUESTED** [review]

File: cpp/examples/nested_types/deduplication.cpp
Comment (divyegala) [reply cpp/examples/nested_types/deduplication.cpp:97]: I think of these functions as analogous to libcudf public APIs, which don't use `auto` return type. It's more readable for users reading this example to know the return type of the function.

File: cpp/examples/nested_types/deduplication.cpp
Comment (vyasr) [reply cpp/examples/nested_types/deduplication.cpp:54]: Ah got it. Then maybe a boolean flag `pool=false` would convey that more simply than a string, which is very flexible. Up to you though.

File: cpp/examples/build.sh
Comment (vyasr) [reply cpp/examples/build.sh:39]: Wow I totally just wrote this like it was a CMake file and not a bash script... Sorry about that.

File: cpp/examples/nested_types/deduplication.cpp
Comment (vyasr) [reply cpp/examples/nested_types/deduplication.cpp]: Good point, agreed.

File: cpp/examples/build.sh
Comment (vyasr) [reply cpp/examples/build.sh:39]: I'm going to implement this now.

File: cpp/examples/nested_types/deduplication.cpp
Comment (vyasr) [reply cpp/examples/nested_types/deduplication.cpp]: Looks like this was done in https://github.com/rapidsai/cudf/pull/13795/commits/f7e96a99716a383f43d6066653f1558af247cdd1

File: cpp/include/cudf/detail/gather.cuh
Comment (vyasr) [reply cpp/include/cudf/detail/gather.cuh]: Looks like this was done in a sequence of commits (e.g. https://github.com/rapidsai/cudf/pull/13795/commits/f2c42be2c8868a23375892921f7f9aae0544568b for range).

File: cpp/examples/nested_types/CMakeLists.txt
Comment (vyasr) [reply cpp/examples/nested_types/CMakeLists.txt]: Done.

File: cpp/examples/nested_types/deduplication.cpp
Comment (vyasr) [reply cpp/examples/nested_types/deduplication.cpp:92]: If we don't force hash aggregations do we still have at least some? As long as the example is demonstrating the all the comparator functionality that we want I think it's totally fine if it isn't the most performant right now (and we've removed timings anyway). If there are improvements to be made to libcudf, we should open a separate issue for that.

**Review by vyasr — APPROVED** [review]

File: cpp/examples/nested_types/deduplication.cpp
Comment (divyegala) [reply cpp/examples/nested_types/deduplication.cpp:92]: I don't get it, what do you mean by "at least have some"?

Comment (bdice) [reply cpp/examples/nested_types/deduplication.cpp:92]: 1. We need to file an issue that makes this use hash-based aggregations without any "forcing"
2. I think Vyas is asking, if you don't add this column to "force" hash aggregations, are _some_ of the aggregations hash-based and others sort-based? My understanding of libcudf's behavior is that if any aggregation is sort-based, all the aggregations fall back to using sort-based implementations. Is that true?

Comment (divyegala) [reply cpp/examples/nested_types/deduplication.cpp:92]: Yes, that's right. If an aggregation only has a sort-based implementation then it will fall back to a sort aggregation.

In this particular case though, a hash-based aggregation is available and it needs to be "forced" due to a limitation.

Here's the issue: https://github.com/rapidsai/cudf/issues/14412

File: cpp/examples/nested_types/deduplication.cpp
Comment (divyegala) [reply cpp/examples/nested_types/deduplication.cpp:64]: https://github.com/rapidsai/cudf/issues/14413

File: cpp/examples/nested_types/deduplication.cpp
Comment (divyegala) [reply cpp/examples/nested_types/deduplication.cpp:184]: So `std::cout` or `CUDF_LOG_INFO`?

File: cpp/examples/nested_types/deduplication.cpp
Comment (vyasr) [reply cpp/examples/nested_types/deduplication.cpp:92]: @bdice are you OK with marking this thread as resolved for now? The issue looks like it has a reasonable plan for how to move forward, but I don't think we should block this PR on getting that enabled.

File: cpp/examples/nested_types/CMakeLists.txt
Comment (vyasr) [reply cpp/examples/nested_types/CMakeLists.txt:16]: `target_compile_features` is the newer option. As for why to prefer it, in this case I don't think it matters too much although there are subtle differences in how it interacts with other CMake variables like `CMAKE_CXX_STANDARD`. It would be more helpful if you wanted it to not be `PRIVATE` but instead `PUBLIC`/`INTERFACE` because compile features will propagate to consuming targets while properties won't.

File: cpp/examples/nested_types/deduplication.cpp
Comment (bdice) [reply cpp/examples/nested_types/deduplication.cpp:184]: Your call. I vote `std::cout` but either is fine.

Comment (vyasr) [reply cpp/examples/nested_types/deduplication.cpp:184]: I think that since none of the other examples currently do this we should skip for now and come back to add in a follow-up if we want. I do think some sort of logging would be nice, but we should take the opportunity to establish a consistent standard of some sort while we still have a small number of examples and it's easy to make them consistent.

File: cpp/examples/nested_types/CMakeLists.txt
Comment (robertmaynard) [reply cpp/examples/nested_types/CMakeLists.txt:16]: The `CXX_STANDARD` property is a request for a compiler level, but can decay to the highest supported level. You need to also set `CXX_STANDARD_REQUIRED` to stop the decaying.

`target_compile_features(deduplication PRIVATE cxx_std_17)` is a hard requirement and errors out instead of decaying.

File: cpp/examples/nested_types/deduplication.cpp
Comment (vyasr) [reply cpp/examples/nested_types/deduplication.cpp:186]: Ah looks like Divye added some. All good then.

File: cpp/examples/nested_types/CMakeLists.txt
Comment (bdice) [reply cpp/examples/nested_types/CMakeLists.txt:16]: Thanks, both of you.

Comment (bdice) [reply cpp/examples/nested_types/CMakeLists.txt:16]: I struggle with knowing what "the newer/better option" is most of the time. 😕

**Review by bdice — APPROVED** [review]

**Review by karthikeyann — APPROVED** [review]
LGTM 👍

**Review by PointKernel — APPROVED** [review]

**Review by ttnghia — APPROVED** [review]

**Review by raydouglass — APPROVED** [review]

**Pull request merged**
